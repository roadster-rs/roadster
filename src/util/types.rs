use std::any::type_name;

/// Get the name of the type with its module prefix stripped.
pub fn simple_type_name<T>() -> String
where
{
    type_name::<T>()
        .split("::")
        .last()
        .unwrap_or(type_name::<T>())
        .to_owned()
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    struct Foo;

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn simple_type_name() {
        let simple_type_name = super::simple_type_name::<Foo>();
        assert_debug_snapshot!(simple_type_name);
    }
}
