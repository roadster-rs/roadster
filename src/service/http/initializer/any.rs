use crate::app::context::AppContext;
use crate::error::RoadsterResult;
use crate::service::http::initializer::Initializer;
use axum::Router;
use axum_core::extract::FromRef;
use typed_builder::TypedBuilder;

#[derive(TypedBuilder)]
pub struct AnyInitializer<S>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
{
    #[builder(setter(into))]
    name: String,
    #[builder(default, setter(strip_option))]
    enabled: Option<bool>,
    #[builder(default, setter(strip_option))]
    priority: Option<i32>,
    #[builder(default, setter(strip_option))]
    stage: Option<Stage>,
    #[builder(setter(transform = |a: impl Fn(Router, &S) -> RoadsterResult<Router> + Send + 'static| to_box_fn(a) ))]
    apply: Box<dyn Fn(Router, &S) -> RoadsterResult<Router> + Send>,
}

#[derive(Default)]
#[non_exhaustive]
pub enum Stage {
    AfterRouter,
    BeforeMiddleware,
    AfterMiddleware,
    #[default]
    BeforeServe,
}

fn to_box_fn<S>(
    p: impl Fn(Router, &S) -> RoadsterResult<Router> + Send + 'static,
) -> Box<dyn Fn(Router, &S) -> RoadsterResult<Router> + Send> {
    Box::new(p)
}

impl<S> Initializer<S> for AnyInitializer<S>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
{
    fn name(&self) -> String {
        self.name.clone()
    }

    fn enabled(&self, state: &S) -> bool {
        let context = AppContext::from_ref(state);
        let config = context
            .config()
            .service
            .http
            .custom
            .initializer
            .custom
            .get(&self.name);
        if let Some(config) = config {
            config.common.enabled(state)
        } else {
            context
                .config()
                .service
                .http
                .custom
                .initializer
                .default_enable
                || self.enabled.unwrap_or_default()
        }
    }

    fn priority(&self, state: &S) -> i32 {
        AppContext::from_ref(state)
            .config()
            .service
            .http
            .custom
            .initializer
            .custom
            .get(&self.name)
            .map(|config| config.common.priority)
            .unwrap_or_else(|| self.priority.unwrap_or_default())
    }

    fn after_router(&self, router: Router, _state: &S) -> RoadsterResult<Router> {
        if let Some(Stage::AfterRouter) = self.stage {
            (self.apply)(router, _state)
        } else {
            Ok(router)
        }
    }

    fn before_middleware(&self, router: Router, _state: &S) -> RoadsterResult<Router> {
        if let Some(Stage::BeforeMiddleware) = self.stage {
            (self.apply)(router, _state)
        } else {
            Ok(router)
        }
    }

    fn after_middleware(&self, router: Router, _state: &S) -> RoadsterResult<Router> {
        if let Some(Stage::AfterMiddleware) = self.stage {
            (self.apply)(router, _state)
        } else {
            Ok(router)
        }
    }

    fn before_serve(&self, router: Router, _state: &S) -> RoadsterResult<Router> {
        if let Some(Stage::BeforeServe) = self.stage {
            (self.apply)(router, _state)
        } else {
            Ok(router)
        }
    }
}
