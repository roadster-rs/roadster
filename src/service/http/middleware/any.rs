use crate::app::context::AppContext;
use crate::error::RoadsterResult;
use crate::service::http::middleware::Middleware;
use axum::Router;
use axum_core::extract::FromRef;
use typed_builder::TypedBuilder;

type ApplyFn<S> = Box<dyn Fn(Router, &S) -> RoadsterResult<Router> + Send>;

/// A [`Middleware`] that can be applied without creating a separate `struct`. Useful to easily
/// apply a middleware that's based on a function, for example.
///
/// # Examples
/// ```rust
/// # use axum::response::Response;
/// # use axum::middleware::Next;
/// # use axum_core::extract::Request;
/// # use tracing::info;
/// # use roadster::service::http::middleware::any::AnyMiddleware;
/// #
/// pub(crate) async fn hello_world_middleware_fn(request: Request, next: Next) -> Response {
///     info!("Running `hello-world` middleware");
///
///     next.run(request).await
/// }
///
/// let middleware = AnyMiddleware::builder()
///     .name("hello-world")
///     .enabled(true)
///     .apply(|router, _state| {
///         Ok(router
///             .layer(axum::middleware::from_fn(hello_world_middleware_fn)))
///     })
///     .build();
/// ```
#[derive(TypedBuilder)]
#[non_exhaustive]
pub struct AnyMiddleware<S>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
{
    #[builder(setter(into))]
    name: String,
    #[builder(default, setter(strip_option))]
    enabled: Option<bool>,
    #[builder(default, setter(strip_option))]
    priority: Option<i32>,
    #[builder(setter(transform = |a: impl Fn(Router, &S) -> RoadsterResult<Router> + Send + 'static| to_box_fn(a) ))]
    apply: ApplyFn<S>,
}

fn to_box_fn<S>(p: impl Fn(Router, &S) -> RoadsterResult<Router> + Send + 'static) -> ApplyFn<S> {
    Box::new(p)
}

impl<S> Middleware<S> for AnyMiddleware<S>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
{
    fn name(&self) -> String {
        self.name.clone()
    }

    fn enabled(&self, state: &S) -> bool {
        let context = AppContext::from_ref(state);
        let config = context
            .config()
            .service
            .http
            .custom
            .middleware
            .custom
            .get(&self.name);
        if let Some(config) = config {
            config.common.enabled(state)
        } else {
            context
                .config()
                .service
                .http
                .custom
                .middleware
                .default_enable
                || self.enabled.unwrap_or_default()
        }
    }

    fn priority(&self, state: &S) -> i32 {
        AppContext::from_ref(state)
            .config()
            .service
            .http
            .custom
            .middleware
            .custom
            .get(&self.name)
            .map(|config| config.common.priority)
            .unwrap_or_else(|| self.priority.unwrap_or_default())
    }

    fn install(&self, router: Router, state: &S) -> RoadsterResult<Router> {
        (self.apply)(router, state)
    }
}
