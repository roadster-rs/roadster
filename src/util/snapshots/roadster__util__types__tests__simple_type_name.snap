---
source: src/util/types.rs
expression: simple_type_name
---
"Foo"
